### ISSUE TITLE
Noticed some Voice Over Issues

### ISSUE THREAD

**Issue by terracoda** [phetsims/john-travoltage#235]
@jessegreenberg, in testing version 31 with
Mac OS Sierra 10.12.3 with Safari 10.0.3

I could use the sim quite well with Voice Over on, but had the following random interruptions while testing the sim.

- [x] Sometimes, electrons are announced one at a time, and sometimes just announced the total amount.	Not sure why this happens.
- [x] I randomly loose focus from time to time. This varies between versions, so it may be something in the sim that is causing the loss of focus??? Not sure.
- [x] Once arm stopped working - just wouldn't move, though focus was there and the non-changing position was being read out.
- [x] Once Leg stopped moving - same as above, but at a different time. Not sure what causes an appendage to stop. Could be a VO bug. VO just seems to konk out sometimes. 
- [x] I hear "Sim sound on." at top of document before anything else. I guess that's Live region alert. Likely, nothing we can do about this, but just thought I would mention it, just in case we can.

Not sure if these items are being tracked elsewhere, and just wanted to jot them down, just in case.

**Comment by jessegreenberg** [comment]
> Sometimes, electrons are announced one at a time, and sometimes just announced the total amount.	Not sure why this happens.

I have noticed this sometimes too, I have been assuming this is just the way VO handles ```aria-live```.  If its an issue we can spend some time investigating.

>I randomly loose focus from time to time. This varies between versions, so it may be something in the sim that is causing the loss of focus??? Not sure.
Once arm stopped working - just wouldn't move, though focus was there and the non-changing position was being read out.
Once Leg stopped moving - same as above, but at a different time. Not sure what causes an appendage to stop. Could be a VO bug. VO just seems to konk out sometimes.

I am on 10.12.1 on my VO testing device, I will update and see if I can reproduce any of these.

> I hear "Sim sound on." at top of document before anything else. I guess that's Live region alert. Likely, nothing we can do about this, but just thought I would mention it, just in case we can.

Good catch! Yes we can fix this.

**Comment by jessegreenberg** [comment]
> I randomly loose focus from time to time.

I am able to see this happen on MacOS 10.12.4 with VO and Safari.  It always happens when the system is stressed, such as when I move the arm quickly around the doorknob. This is a performance issue. VO says Safari is "busy" after focus is lost.

**Comment by jessegreenberg** [comment]
While playing with the sim, I ran into a case where the appendages stopped moving.   I noticed that when this happens VoiceOver says "Quick nav on", and quick nav is turned on by pressing left and right arrow keys at the same time.  When "Quick nav" is on, the sliders don't work.

**Comment by jessegreenberg** [comment]
Looking at this again, I removed all electrons and the performance is much better in Safari, but the focus will still leave the appendages.  The focus will either go to the H1, or a random number.

Sim.js has a "heartbeat div" that gets a random value every 10 seconds or so.  I wonder if the VO moves focus when the value changes for some reason??

**Comment by jessegreenberg** [comment]
Yes, that must be it, when the number changes focus leaves the appendages.

**Comment by terracoda** [comment]
@jessegreenberg, yes, I heard "Quick Nav on" when using the arrow keys quickly. Sorry, I didn't report that.

**Comment by jessegreenberg** [comment]
Regarding 

> I randomly loose focus from time to time. This varies between versions, so it may be something in the sim that is causing the loss of focus??? Not sure.

I think I figured this out. The discharge alert goes through an element with ```aria-live=polite``` and ```role-status```.  When I remove the status role on the element focus never leaves the arm or leg.

**Comment by jessegreenberg** [comment]
While using ```role=status``` the focus left the arm and leg 8 times in one minute while I was building charge and discharging with VO on.

When using an aria alert without ```role=status``` the focus left the arm and leg 0 times in 2 minutes and 20 seconds.

@jobara is it OK if we send alerts through the alert element that does not use ```role=status```?

**Comment by jessegreenberg** [comment]
Ah, the "heartbeat div" was commented out in my local copy.  When that is added back into the document focus still leaves the sliders spontaneously.

**Comment by terracoda** [comment]
Role `status` was slightly updated in Aria 1.1, but I don't know exactly what was updated, also do not know the level of support Browser/AT combos.

In all the things I read, the element with the `status` role, must not receive focus. @jessegreenberg, is it possible that the element is accidentally getting focus, which makes the focus get lost?

Also, it is possible that the leg and arm may need to use the aria-controls attribute because they control when charges are added and when they get discharged. Is the `aria-controls` attribute being used? If not, it might help.

But maybe it is teh heartbeat div that is causing the problem.

I was looking at these resources:
Thread on aim web-list
http://webaim.org/discussion/mail_thread?thread=6059

MDN Doc
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_status_role

Actual specification:
https://www.w3.org/TR/wai-aria-1.1/
https://www.w3.org/TR/wai-aria-1.1/#status

**Comment by terracoda** [comment]
@jessegreenberg, in this 2014 article, Léonie Watson writes
> the most widely supported combination is to use `aria-live=”assertive”` with `role=”status” `

In the comments, however, James Craig mentions that
> if you put a timeout on that live region update, it works.

I think by "it" Craig is saying that the politeness level works.

There's no mention of focus issues.

**Comment by jessegreenberg** [comment]
>  is it possible that the element is accidentally getting focus, which makes the focus get lost?

@terracoda I don't think so, all ```aria-live``` elements are out of the focus order, and we never focus them.  The problem is that VO+Safari moves the virtual cursor onto the the element with role=status occasionally, I am not sure if this is a bug or an intended feature.

> Is the aria-controls attribute being used?

Yes it is.

> I think by "it" Craig is saying that the politeness levels work with a set timeout.

We found this too, our alert system now uses a timeout by default.

**Comment by jessegreenberg** [comment]
The heartbeat div seems to interfere with focus most consistently after charges are picked up.  Possibly because an ```aria-live``` region was updated?

The heartbeat div will also interfere with focus if I move VO focus into the dev tools and then back into the sim.

**Comment by jessegreenberg** [comment]
If I add the ```hidden```` attribute to the div, the problem seems to go away.  I am not sure if that is acceptable though or if it will totally remove the impact of the div.

**Comment by jessegreenberg** [comment]
The the div hidden, focus did not leave the arm once in 3 minutes 20 seconds.

**Comment by terracoda** [comment]
@jessegreenberg, the number of electrons is updated in the summary, which is not hidden. Thus, the information is still in the PDOM if the user needs it, but hasn't heard it.

Will the alerts for "Sim sound on." and "Sim sound muted." also be` hidden` using the same technique? I could still hear them in the latest version I tested.

**Comment by jessegreenberg** [comment]
> Will the alerts for "Sim sound on." and "Sim sound muted." also behidden using the same technique? I could still hear them in the latest version I tested.

No, they will not be hidden. Also, ```hidden``` wont work for reasons described in https://github.com/phetsims/joist/issues/140, but I think we found another workaround.

**Comment by terracoda** [comment]
Ok :-)

**Comment by jessegreenberg** [comment]
phetsims/joist#140 is fixed, but ```role=status``` is still also causing focus to leave the arm and leg.  It will take a combination of both to fix the focus issues.

**Comment by jessegreenberg** [comment]
The above commit prevents the alert from being announced when the sound button is created.  Everything should be fixed except for this:

> Sometimes, electrons are announced one at a time, and sometimes just announced the total amount. Not sure why this happens.

And I think this is just the way ```aria-live=polite``` works.

@terracoda thanks for this list, I think VoiceOver is working much better now. If you notice these or other issues please let me know.

**Comment by terracoda** [comment]
Awesome @jessegreenberg! And I think you are right about `aria-live=polite`.
